intermediate_source/TP_tutorial.rst

```diff
@@ -0,0 +1,282 @@
+
+
+It performs ``w1`` and ``w3`` matmuls concurrently and followed by a w2 matmul with the result of the combined w1/w3 linear projection results. This means we could use the idea from the Tensor Parallelism paper to shard the w1/w3 Linear layers in the colwise fashion and shard the w2 Linear layer in the rowwise fashion, so that there iss only one allreduce communication happening at the end of all the three layers. With the PyTorch native Tensor Parallel, we can simply create a parallelize_plan for the FeedForward layer like below:
+
+.. code-block:: python
+
+    from torch.distributed.tensor.parallel import ColwiseParallel, RowwiseParallel, parallelize_module
+
+    tp_plan = {
+        # by default ColwiseParallel input layouts is replicated
+        # and RowwiseParallel output layouts is replicated
+        "feed_foward.w1": ColwiseParallel(),
+        "feed_forward.w2": RowwiseParallel(),
+        "feed_forward.w3": ColwiseParallel(),
+    }
+
+
+That's simply how we configure the shardings for the FeedForward layer using the PyTorch Tensor Parallel APIs. Note that users would only need to specify how to shard the individual layers and the communications (i.e. allreduce) will happen under the hood.
+
+Next for the Attention Layer, it consists of ``wq, wk, wv`` Linear layers to project input to q/k/v, and then it performs attention and output projection with the “wo” Linear layer. Tensor Parallelism here intended to perform colwise sharding for the q/k/v projection and rowwise sharding for the wo linear projection. So we can add the Attention plan to the tp_plan that we just drafted up:
+
+.. code-block:: python
+
+    tp_plan = {
+        # by default ColwiseParallel input layouts is replicated
+        # and RowwiseParallel output layouts is replicated
+        "attention.wq": ColwiseParallel(),
+        "attention.wk": ColwiseParallel(),
+        "attention.wv": ColwiseParallel(),
+        "attention.wo": RowwiseParallel(),
+        "feed_foward.w1": ColwiseParallel(),
+        "feed_forward.w2": RowwiseParallel(),
+        "feed_forward.w3": ColwiseParallel(),
+    }
+
+
+This is almost the ``tp_plan`` we need to apply Tensor Parallelism to the llama model, however one thing we should be aware is that when sharding the linear layer colwise, the output of the linear layers would become sharded on the last tensor dimension, and the rowwise sharding linear layer directly accepts an input that shards on the last dimension. If there’re any more tensor operations (i.e. view operations) between the colwise linear and the rowwise linear, we would need to adjust the relevant shape related ops to sharded shape.
+
+For the Llama model, in the attention layer there are couple of view operations that are shape related, the main source is that after colwise parallel for wq/wk/wv linear layers, the activation is sharded on the num_heads dimension, so we would need to adjust the num_heads to local num_heads.
```

Comment:
```suggestion
For the Llama model, in the attention layer there are couple of view operations that are shape related, the main source is that after column-wise parallel for ``wq``/``wk``/``wv`` linear layers, the activation is sharded on the ``num_heads`` dimension, so we would need to adjust the ``num_heads`` to local ``num_heads``.
```

Reply:
there are couple of -> there are **a** couple of

Reply:
Please break this into two sentences.